wallet-client/WalletClient.hs

```diff
@@ -0,0 +1,223 @@
+
+
+data TraceTx tx reject
+  = TraceTxAccept tx
+  | TraceTxReject tx reject
+  deriving Show
+
+instance (Condense tx, Condense reject) => Condense (TraceTx tx reject) where 
+    condense (TraceTxAccept tx)        = "TxAccept "   <> condense tx
+    condense (TraceTxReject tx reject) = "TxRejected " <> condense tx
+                                                       <> " "
+                                                       <> condense reject
+
+
+-- | A 'LocalTxSubmissionClient' that submits transactions reading them from
+-- a 'TMVar'.  A real implementation should use a better synchronisation
+-- primitive.  This demo creates and empty 'TMVar' in
+-- 'muxLocalInitiatorNetworkApplication' above and never fills it with a tx.
+--
+txSubmissionClient
+  :: forall tx reject m.
+     ( Monad    m
+     , MonadSTM m
+     )
+  => Tracer m (TraceTx tx reject)
+  -> TMVar m tx
+  -> m (LocalTxSubmissionClient tx reject m Void)
+txSubmissionClient tracer txv = do
+    tx <- atomically $ readTMVar txv
+    pure $ SendMsgSubmitTx tx $ \mbreject -> do
+      case mbreject of
+        Nothing -> traceWith tracer (TraceTxAccept tx)
+        Just r  -> traceWith tracer (TraceTxReject tx r)
+      txSubmissionClient tracer txv
+
+
+-- | 'ChainSyncClient' which traces received blocks and ignores when it
+-- receives a request to rollbackwar.  A real wallet client should:
+--
+--  * at startup send the list of points of the chain to help synchronise with
+--    the node;
+--  * update its state when the client receives next block or is requested to
+--    rollback, see 'clientStNext' below.
+--
+chainSyncClient
+  :: forall blk m. MonadTimer m
+  => Tracer m blk
+  -> ChainSyncClient blk (Point blk) m Void
+chainSyncClient tracer = ChainSyncClient $ pure $
+    -- Notify the core node about the our latest points at which we are
+    -- synchronised.  This client is not persistent and thus it just
+    -- synchronises from the genesis block.  A real implementation should send
+    -- a list of points up to a point which is k blocks deep.
+    SendMsgFindIntersect
+      [genesisPoint]
+      ClientStIntersect {
+        recvMsgIntersectImproved  = \_ _ -> ChainSyncClient (pure clientStIdle), 
+        recvMsgIntersectUnchanged = \  _ -> ChainSyncClient (pure clientStIdle)
+      }
+  where
+    clientStIdle :: ClientStIdle blk (Point blk) m Void
+    clientStIdle = 
+      SendMsgRequestNext clientStNext (pure clientStNext)
+
+    clientStNext :: ClientStNext blk (Point blk) m Void
+    clientStNext = ClientStNext {
+        recvMsgRollForward = \blk _tip -> ChainSyncClient $ do
+          -- trace the next block, the core note tip point is '_tip'.
+          traceWith tracer blk
+          pure clientStIdle
+      , recvMsgRollBackward = \_point _tip -> ChainSyncClient $ do
+          -- we are requested to roll backward to point '_point', the core
+          -- node's chain's tip is '_tip'.
+          pure clientStIdle
```

Comment:
Should trace here too.

Reply:
I will actually remove the trace.  I'll add a comment how to trace all network messages.